### PR TITLE
importer: add ImportEpoch table descriptor field

### PR DESCRIPTION
**Pull request by msbutler**
This patch introduces the new ImportEpoch table descriptor field which
provides a count for the number of IMPORT INTO jobs that have been attempted on
the table after it already had data, including any in-progress jobs.

This work is apart of a larger project to roll back imports without MVCC
timestamps. In a future pr (https://github.com/cockroachdb/cockroach/pull/85138), this epoch will get written to the
imported MVCCValueHeader metadata, which in turn allows import rollbacks to
use the epoch to determine which keys to delete in the table.

We considered binding the import job id to the table descriptor, and in turn
to each imported KV's metadata, however, the expected proto encoded size of a
job ID was between 7 - 12 bytes, which seemed uncessary. Meanwhile the 4 byte
uint32 epoch will encode to 1 byte for the first 256 IMPORT INTO jobs on the
table. 2,147,483,648 IMPORT INTO jobs must occur on the table before we worry
about integer overflow -- a problem for a future release :D.

Informs #76722

Release note: none

### CODE REVIEW

**Comment by msbutler** [comment]
I realize the unit test gets called within the bulk data driven driver. Once I address the failures in https://github.com/cockroachdb/cockroach/pull/85637, I'll move the unit test to `pkg/ccl/importccl`.

**Review by erikgrinaker — APPROVED** [review]
Looks good at a high level, but this is sufficiently far away from KV that someone more familiar with it should review this too.

I'm assuming this will piggyback on the version gate in #85138, so there are no further cross-version concerns here?

**Comment by msbutler** [comment]
Note: i think I need to rework this PR such that two import epoch's are stored in the table descriptor. Will update later today.

**Comment by dt** [comment]
NextImportEpoch -- counter that increments every time to made sure epochs are unique
CurrentImportEpoch -- value of the epoch currently being used to import, or zero if not currently importing (i.e. only ever non-zero when table is offline and was taken from online to offline because it existed prior to import)

**Comment by msbutler** [comment]
- added the cluster version to this PR
- @dt I didn't quite go with your suggestion, but my approach achieves the same purpose.

**Comment by msbutler** [comment]
sorry for the noise -- RFAL!

File: pkg/sql/catalog/descpb/structured.proto [hunk 1]
```diff
@@ -1180,7 +1180,23 @@ message TableDescriptor {
   // this table, in which case the global setting is used.
   optional bool forecast_stats = 52 [(gogoproto.nullable) = true, (gogoproto.customname) = "ForecastStats"];
 
-  // Next ID: 54
+  // ImportIntoEpoch is the count of IMPORT INTO jobs that have been attempted on this
```
Comment (erikgrinaker) [inline pkg/sql/catalog/descpb/structured.proto:1183]:
```suggestion
  // ImportEpoch is the count of IMPORT INTO jobs that have been attempted on this
```

File: pkg/sql/catalog/descpb/structured.proto [hunk 2]
```diff
@@ -1180,7 +1180,23 @@ message TableDescriptor {
...
+  // table once it __already had data__. The ImportEpoch can include an in progress import, as the
+  // field gets incremented while preparing the table for ingestion.
+  optional uint32 import_epoch = 54 [(gogoproto.nullable) = false, (gogoproto.customname) = "ImportEpoch"];
+
+  // ImportType indicates the kind of import currently occurring on the table.
```
Comment (erikgrinaker) [inline pkg/sql/catalog/descpb/structured.proto:1188]: This could explain the differences. Why is `IMPORT` a separate thing?

**Review by erikgrinaker — COMMENTED** [review]
I'll still defer to Bulk/Schema here, but this seems fine as far as I can tell.

I'm assuming we already have some sort of mutual exclusion protocol here to prevent two import jobs from clobbering each others import state?

**Comment by msbutler** [comment]
closing in favor of https://github.com/cockroachdb/cockroach/pull/85852

**Pull request closed without merging**